### PR TITLE
[Beam:Vod] Add new extractor

### PR DESCRIPTION
**Pull request by mikf**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This is a new extractor for VODs from the https://beam.pro/ streaming service.

### CODE REVIEW

File: youtube_dl/extractor/beampro.py [hunk 1]
```diff
@@ -71,3 +73,93 @@ def _real_extract(self, url):
...
+            'duration': float_or_none(vod_info.get('duration')),
+            'thumbnail': thumbnail_url,
+            'timestamp': parse_iso8601(vod_info.get('createdAt')),
+            'uploader': chan.get('token') or try_get(
+                chan, lambda x: x['user']['username'], compat_str),
+            'uploader_id': compat_str(user_id) if user_id else None,
+            'age_limit': BeamProLiveIE._RATINGS.get(chan.get('audience')),
+            'view_count': int_or_none(vod_info.get('viewsTotal')),
+            'formats': formats,
+        }
```
Comment (dstftw) [inline youtube_dl/extractor/beampro.py]: 151-165 code duplication.

File: youtube_dl/extractor/beampro.py [hunk 2]
```diff
@@ -6,15 +6,17 @@
...
     parse_iso8601,
     try_get,
+    urljoin,
 )
 
 
 class BeamProLiveIE(InfoExtractor):
     IE_NAME = 'Beam:live'
-    _VALID_URL = r'https?://(?:\w+\.)?beam\.pro/(?P<id>[^/?#&]+)'
+    _VALID_URL = r'https?://(?:\w+\.)?beam\.pro/(?P<id>[^/?#&]+)$'
```
Comment (dstftw) [inline youtube_dl/extractor/beampro.py]: Should match `http://www.beam.pro/niterhayven/` and `http://www.beam.pro/niterhayven?abc=123`.

File: youtube_dl/extractor/beampro.py (hunk 1 above)
Comment (mikf) [inline youtube_dl/extractor/beampro.py]: I'm assuming you mean lines 151-165 contain duplicate code from the BeamProLiveIE class.
Is it OK if I just create a common base class and put all the duplicate code in there?

File: youtube_dl/extractor/beampro.py (hunk 2 above)
Comment (mikf) [inline youtube_dl/extractor/beampro.py]: The issue here is that `http://www.beam.pro/niterhayven?vod=123`, which is a VOD URL, would also get matched by this. I tried to fix this in ccc035c1fb2b6080b4dab9e2cb0b8273f734c1cc by reordering the import of those extractors, but I don't know if this acceptable.

Comment (dstftw) [inline youtube_dl/extractor/beampro.py]: Override `suitable`.

**Comment by mikf** [comment]
The requested changes should have all been addressed.
Are there any other issues left or is there anything else that I should do?

On another note: Beam VODs are only available for 2 weeks after creation, so any test is going to fail after that time. Is there a preferred way of handling cases like that?

**Pull request closed without merging**